beacon-chain/rpc/prysm/v1alpha1/validator/proposer_altair.go

```diff
@@ -107,3 +119,89 @@ func (vs *Server) getSyncAggregate(ctx context.Context, slot primitives.Slot, ro
 		SyncCommitteeSignature: syncSigBytes[:],
 	}, nil
 }
+
+func (vs *Server) aggregatedSyncCommitteeMessages(ctx context.Context, slot primitives.Slot, root [32]byte) ([]*ethpb.SyncCommitteeContribution, error) {
+	subcommitteeCount := params.BeaconConfig().SyncCommitteeSubnetCount
+	subcommitteeSize := params.BeaconConfig().SyncCommitteeSize / subcommitteeCount
+	sigsPerSubcommittee := make([][][]byte, subcommitteeCount)
+	bitsPerSubcommittee := make([]bitfield.Bitfield, subcommitteeCount)
+	for i := uint64(0); i < subcommitteeCount; i++ {
+		sigsPerSubcommittee[i] = make([][]byte, 0, subcommitteeSize)
+		bitsPerSubcommittee[i] = ethpb.NewSyncCommitteeAggregationBits()
+	}
+
+	// Get committee position(s) for each message's validator index.
+	scMessages, err := vs.SyncCommitteePool.SyncCommitteeMessages(slot)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get sync committee messages")
+	}
+	messageIndices := make([]primitives.ValidatorIndex, 0, len(scMessages))
+	messageSigs := make([][]byte, 0, len(scMessages))
+	for _, msg := range scMessages {
+		if bytes.Equal(root[:], msg.BlockRoot) {
+			messageIndices = append(messageIndices, msg.ValidatorIndex)
+			messageSigs = append(messageSigs, msg.Signature)
+		}
+	}
+	headSyncCommitteeIndices := make([][]primitives.CommitteeIndex, len(messageIndices))
+	for i, vi := range messageIndices {
+		headSyncCommitteeIndices[i], err = vs.HeadFetcher.HeadSyncCommitteeIndices(ctx, vi, slot)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not get head sync committee indices for validator index %d", vi)
+		}
+	}
+
+	// Based on committee position(s), set the appropriate subcommittee bit and signature.
+	for i, indices := range headSyncCommitteeIndices {
+		for _, index := range indices {
+			k := uint64(index)
+			subnetIndex := k / subcommitteeSize
+			indexMod := k % subcommitteeSize
+			if !bitsPerSubcommittee[subnetIndex].BitAt(indexMod) {
+				bitsPerSubcommittee[subnetIndex].SetBitAt(indexMod, true)
+				sigsPerSubcommittee[subnetIndex] = append(sigsPerSubcommittee[subnetIndex], messageSigs[i])
+			}
+		}
+	}
+
+	// Aggregate. To improve performance, we aggregate messages for each subcommittee in parallel.
+	// Using a channel is safer than a slice because appending to a slice is not thread-safe.
+	contributionsChan := make(chan *ethpb.SyncCommitteeContribution, subcommitteeCount)
+	var wg sync.WaitGroup
+	wg.Add(int(subcommitteeCount)) // lint:ignore uintcast -- Subcommittee count will never exceed maxint
```

Comment:
I think this parallelization may be premature optimization. There are only 512 sync committee members and while a unit bench may show in isolation that parallelizing this we get improvement. In production we already have a lot of routines working in parallel at block production time and it's not clear to me that this extra overhead here would be helpful.